Saving and loading of keys and ciphertexts

CarlosMoralesSal opened the pull request:
Initial commit to create save and load methods for Public and Secret Key

CarlosMoralesSal committed:
Initial Commit

CarlosMoralesSal committed:
Remove some comments

CarlosMoralesSal committed:
Remove comment from logistic example

CarlosMoralesSal commented:
Remove another comment

CarlosMoralesSal committed:
remove load methods and fixing save outputs

CarlosMoralesSal committed:
Remove several files and cleaning code

mortendahl commented on Makefile:
```diff
@@ -103,7 +103,7 @@ ifeq (,$(shell grep -e $(VERSION) setup.py))
 
 twine:
-	pip install --upgrade setuptools wheel twine
+	python3.7 -m pip install --upgrade setuptools wheel twine
```
what was the need for this?

CarlosMoralesSal replied:
Fixed on next commit

mortendahl commented on examples/matmul.py:
```diff
@@ -24,3 +24,5 @@
 with tf.Session() as sess:
     expected, actual = sess.run([c, c_encrypted])
     np.testing.assert_almost_equal(actual, expected, decimal=3)
+    print(expected)
```
leftover?

CarlosMoralesSal replied:
Removed and commited

mortendahl commented on tf_seal/cc/kernels/seal_kernels.cc:
```diff
@@ -79,6 +79,58 @@ Status LookupOrCreateWrapper(OpKernelContext* ctx,
+
+
+  void Compute(OpKernelContext* ctx) override {
+      const PublicKeysVariant* key = nullptr;
```
small formatting issues here and below

CarlosMoralesSal replied:
Done

mortendahl commented on tf_seal/cc/kernels/seal_kernels.cc:
```diff
@@ -79,6 +79,58 @@ Status LookupOrCreateWrapper(OpKernelContext* ctx,
+    OP_REQUIRES_OK(ctx, ctx->allocate_output(0, TensorShape{}, &out0));
+    seal::PublicKey publicKey(key->public_key);
+    std::filebuf fb;
+    fb.open("public_key", std::ios::out);
```
we should make the filename an attribute or ideally an input (string scalar)

CarlosMoralesSal replied:
@mortendahl , In this case, If I understood, the seal_ops, I must declared on this way

    REGISTER_OP("SealSavePublickey")
       .Input("nameSavePK: string")
       .Input("pub_key: variant")    
       .Output("out:variant")
       .SetIsStateful();

Then on kernel I must do this?
```
const Tensor& input_tensor=ctx->input(0);
      std::string input = input_tensor.flat<std::string>();
```

Or how?. I see on the last line I have a mistake. Thanks for all

mortendahl replied:
see eg reply on slack: I think you may have to use a slightly more complex expression, eg like so: https://github.com/tf-encrypted/tf-seal/blob/master/tf_seal/cc/kernels/seal_kernels.cc#L44 (since you are working with tensors, even if flattened)

CarlosMoralesSal replied:
Done. @mortendahl , I have pushed it

mortendahl commented on tf_seal/cc/kernels/seal_kernels.cc:
```diff
@@ -79,6 +79,58 @@ Status LookupOrCreateWrapper(OpKernelContext* ctx,
                                                      });
 }
 
+class SealSavePbOp : public OpKernel {
```
I think we should be more explicit re naming, say `SealSavePublickeyOp`

CarlosMoralesSal replied:
Done

mortendahl commented on tf_seal/cc/kernels/seal_kernels.cc:
```diff
@@ -79,6 +79,58 @@ Status LookupOrCreateWrapper(OpKernelContext* ctx,
+    std::filebuf fb;
+    fb.open("secret_key", std::ios::out);
+    std::ostream pubk(&fb);
+    secretKey.save(pubk);
```
it's probably a good idea to explicitly close the file when we're done with it; ideally adding a bit of error handling as well

CarlosMoralesSal replied:
Done in next commit

CarlosMoralesSal committed:
Removing more files and modifying python version on configure

mortendahl reviewed (commented):
great first steps! 💪

CarlosMoralesSal committed:
Removing pip version from Makefile

CarlosMoralesSal committed:
Adding close method to files and change names to new methods

mortendahl commented on tf_seal/cc/ops/seal_ops.cc:
```diff
@@ -72,3 +72,18 @@ REGISTER_OP("SealPolyEval")
+    .Output("out:variant")
+    .SetIsStateful();
+
+REGISTER_OP("SaveSecretkey")
```
`SealSaveSecretkey` (to match `SealSavePublickey`)?

CarlosMoralesSal replied:
I want to save both, public and secret. I saw a small typo on kernel

mortendahl replied:
oh, I mean that you've left out the `Seal` prefix on this one

CarlosMoralesSal replied:
Done

mortendahl commented on tf_seal/cc/ops/seal_ops.cc:
```diff
@@ -72,3 +72,18 @@ REGISTER_OP("SealPolyEval")
+    .Output("out:variant")
+    .SetIsStateful();
+
+REGISTER_OP("SaveCipherText")
```
same: `SealSaveCiphertext`?

CarlosMoralesSal replied:
After Public and Secret I want to save the Ciphertext

CarlosMoralesSal replied:
Done

CarlosMoralesSal committed:
input string for filenames

CarlosMoralesSal commented:
New changes for filenames

CarlosMoralesSal committed:
Changes on OPs for SaveMethod

CarlosMoralesSal committed:
Uploading load methods

CarlosMoralesSal committed:
New changes on kernels

CarlosMoralesSal commented:
I made some changes yesterday in save and load methods

CarlosMoralesSal commented:
@mortendahl , I have a doubt, on the _seal_ops.so when you load it?, how to know the name of the methods which comes from the ops.cc?, I don't see the trace, so probably my methods aren't called because of this.

CarlosMoralesSal committed:
new changes on kernels and remove comments

CarlosMoralesSal committed:
Upload dummy OP to check if it works and modifying seal_ops.py

mortendahl commented on tf_seal/cc/ops/seal_ops.cc:
```diff
@@ -9,6 +9,14 @@ REGISTER_OP("SealKeyGen")
     .Output("sec_key: variant")
     .SetIsStateful();
 
+REGISTER_OP("SealDummy")
```
What is this guy for?

CarlosMoralesSal replied:
I proced to clean it

CarlosMoralesSal replied:
Done it

mortendahl committed:
cleanup

mortendahl commented on examples/matmul.py:
```diff
@@ -5,8 +5,8 @@
 # sample inputs in the form of tf.Tensors
-a = tf.random.normal(shape=(2, 3), dtype=tf.float32)
-b = tf.random.normal(shape=(2, 3), dtype=tf.float32)
+a = tf.random.normal(shape=(20, 20), dtype=tf.float32)
```
Did you have a particular reason for increasing these?

CarlosMoralesSal replied:
Just testing, I come back to put it on the same number as before

CarlosMoralesSal replied:
Done it

mortendahl committed:
save publickey with test

mortendahl committed:
load publickey with test

mortendahl committed:
formatting

CarlosMoralesSal committed:
Removing Dummy OP

CarlosMoralesSal committed:
Modifying logistic_regression example

CarlosMoralesSal committed:
removing from ops dummy

CarlosMoralesSal committed:
Removing from matmul example

CarlosMoralesSal committed:
Implement Save Secretkey

CarlosMoralesSal committed:
change permissions

CarlosMoralesSal committed:
New changes for save ciphertexts and load it

mortendahl commented on tf_seal/cc/ops/seal_ops.cc:
```diff
@@ -72,3 +72,43 @@ REGISTER_OP("SealPolyEval")
+    .Input("pub_key: variant") 
+    .SetIsStateful();
+
+REGISTER_OP("SealSaveCipher")
```
What is the difference between `SealSaveCipher` and `SealSaveCiphertext`?

CarlosMoralesSal replied:
I have remove the second one.

mortendahl commented on tf_seal/cc/kernels/seal_kernels.cc:
```diff
@@ -79,6 +79,122 @@ Status LookupOrCreateWrapper(OpKernelContext* ctx,
+      file.close();
+    } else {
+        std::cout << "Error opening file!" << std::endl;
+        exit(1);
```
Perhaps it's actually better to replace the `if` branch with a `OP_REQUIRES(ctx, file.is_open(), errors::InvalidArgument("Could not open file ", filename));` which is a cleaner way to exit

mortendahl reviewed (commented):
For saving/loading ciphertexts I think it might be better to switch to protobuf files, storing each SEAL ciphertext as a binary blob (`bytes`) with additional fields for the metadata (ie rows and cols). The message might look something like this:

```protobuf
message SealCiphertext {
  int64 rows = 1;
  int64 cols = 2;
  repeated bytes raw_ciphertexts = 3;
}
```

There are some [official concerns about using protobuf for large data](https://developers.google.com/protocol-buffers/docs/techniques#large-data), but as far as I can tell this is [only a problem if you need random access](https://stackoverflow.com/questions/47564437/why-protobuf-is-bad-for-large-data-structures), which we don't.

I believe we need to keep the `CipherTensor` class around, so one implementation approach would be to add `Save` and `Load` methods to `CipherTensor` that internally uses protobuf messages to store the metadata and the converted SEAL ciphertexts. Converting the SEAL ciphertexts could probably be done using [`save`](https://github.com/microsoft/SEAL/blob/master/native/src/seal/ciphertext.h#L505-L517)/[`load`](https://github.com/microsoft/SEAL/blob/master/native/src/seal/ciphertext.h#L526-L542) together with [`ostringstream`](http://www.cplusplus.com/reference/sstream/ostringstream/)/[`istringstream`](http://www.cplusplus.com/reference/sstream/istringstream/). The protobuf messages can be saved/loaded to disk using the generated `SerializeToOstream`/`ParseFromIstream`.

See also:
- [C++ protobuf tutorial](https://developers.google.com/protocol-buffers/docs/cpptutorial); note that we should probably use syntax version 3.
- [Protobuf language guide](https://developers.google.com/protocol-buffers/docs/proto3).
- [cereal](http://uscilab.github.io/cereal/) is an alternative for serialization, but long term I think protobuf is better since everything else in TensorFlow is using it.

cc @justin1121: any thoughts?

CarlosMoralesSal committed:
Add proto file

mortendahl committed:
compile protobuf with bazel

mortendahl committed:
additional protobuf messages

mortendahl committed:
increate bazel test timeouts

`long` seems a bit too much, `moderate` (the default) should be enough; see https://docs.bazel.build/versions/master/test-encyclopedia.html

mortendahl committed:
test in separate circle ci task

CarlosMoralesSal committed:
working with ostringstream to save

mortendahl committed:
tiny cleanup

mortendahl committed:
save cipher as EncryptedTensor protobuf

mortendahl committed:
load cipher from EncryptedTensor protobuf

mortendahl committed:
quick fix for test_save_load_cipher

justin1121 commented:
> cc @justin1121: any thoughts?

@mortendahl seems reasonable to me to store these things in a protobuf file. Are there other standards for storing keys that we could perhaps look into?

@CarlosMoralesSal what's the status of this PR? No longer actively looking at it?

CarlosMoralesSal commented:
@justin1121 I want to do some changes first
